src/core/org/apache/jmeter/gui/UndoHistory.java

```diff
@@ -160,63 +133,46 @@ public void add(JMeterTreeModel treeModel, String comment) {
         // first clone to not convert original tree
         tree = (HashTree) tree.getTree(tree.getArray()[0]).clone();
 
-        position++;
-        while (history.size() > position) {
-            if (log.isDebugEnabled()) {
-                log.debug("Removing further record, position: {}, size: {}", position, history.size());
-            }
-            history.remove(history.size() - 1);
-        }
-
         // cloning is required because we need to immute stored data
         HashTree copy = UndoCommand.convertAndCloneSubTree(tree);
 
-        history.add(new UndoHistoryItem(copy, comment));
+        GuiPackage guiPackage = GuiPackage.getInstance();
+        //or maybe a Boolean?
+        boolean dirty = guiPackage != null ? guiPackage.isDirty() : false;
```

Comment:
I investigated further code and behaviour, I think issue in incorrect dirty detection is due to the hypothesis made on GuiPackage#isDirty(), it appears it is not up to date when added to UndoHistory.